keps/sig-auth/267-kubelet-server-certificate-bootstrap-rotation/README.md

```diff
@@ -0,0 +1,292 @@
+- request serving certificates from the cluster server, via the CSR API
+
+## Motivation
+
+The client certificate provided by TLS bootstrapping is signed, by default, for client auth only, and thus cannot be used as serving certificates, or server auth.
+
+This proposal covers a process for generating a key locally and then issuing a
+Certificate Signing Request to the cluster API server to get an associated
+certificate signed by the cluster Certificate Authority. Also, as certificates
+approach expiration, the same mechanism will be used to request an updated
+certificate.
+
+### Goals
+
+- Allow Nodes/Kubelet to request a certificate to an external authority instead of using self signed certificates during bootstrap.
+- Allow Nodes/Kubelet to request renewals of their certificates to an external authority.
+
+
+## Design Details
+
+Kubernetes v1.8 and higher kubelet implements features for enabling rotation of its client and/or serving certificates.
+
+You can configure the kubelet to rotate its client certificates by creating new CSRs as its existing credentials expire. To enable this feature, user can enable the rotateCertificates field of kubelet configuration file.
+
+The CSR approving controllers implemented in core Kubernetes do not approve node serving certificates for security reasons. To use RotateKubeletServerCertificate operators need to run a custom approving controller, or manually approve the serving certificate requests.
+
+A deployment-specific approval process for kubelet serving certificates should typically only approve CSRs which:
+
+1. are requested by nodes (ensure the spec.username field is of the form system:node:<nodeName> and spec.groups contains system:nodes)
+
+2. request usages for a serving certificate (ensure spec.usages contains server auth, optionally contains digital signature and key encipherment, and contains no other usages)
+
+3. only have IP and DNS subjectAltNames that belong to the requesting node, and have no URI and Email subjectAltNames (parse the x509 Certificate Signing Request in spec.request to verify subjectAltNames)
+
+### Test Plan
+
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+
+##### Unit tests
+
+- `k8s.io/kubernetes/pkg/kubelet/certificate/kubelet.go`: `2024-09-12` - `57.7`
+- k8s.io/kubernetes/pkg/kubelet/certificate/transport.go: `2024-09-12` - `64.1`
+- k8s.io/kubernetes/pkg/kubelet/certificate/bootstrap/bootstrap.go: `2024-09-12` - `50`
+
+- `<package>`: `<date>` - `<test coverage>`
+
+##### Integration tests
+
+##### e2e tests
+
+Any cluster running e2e tests with the feature enabled will be exercising the feature.
+
+A job using the built-in CSR approver will be added exercising all the Conformance e2e tests.
```

Comment:
do we need to run all tests? Once kubelet bootstrapped, there is nothing new when we run conformance. Would it be best to concentrate on covering edge cases like various failures handling and certs rotation?

Reply:
after 20 releases what regressions or errors we plan to uncover?

Reply:
I mean, the test is to run existing jobs with the builtin cloudprovider, everything should pass

Reply:
we can setup a time to rotate certificates very low, so the job guarantee that the certificates are rotated during the execution of the tests